src/GenericList.hpp

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#include <list>
+#include <optional>
+#include <utility>
+
+#include <range/v3/algorithm/find.hpp>
+
+// A generic, modification-while-iterating-over safe list of objects.
+// This makes it safe to write code like:
+// for (auto *obj : things) { things.remove(obj); }
+//
+// Objects are stored in a doubly-linked list.
+// Each doubly-linked node has a reference count: removing an item with a non-zero reference doesn't actually remove it
+// from the list, instead storing a "tombstone" at that list location. Iteration skips tombstones (and cleans them up as
+// it finds it can).
+
+template <typename ElemT, typename GenListT, typename It>
+class Iter;
+
+template <typename T, typename OptType = std::optional<T>>
+class GenericList {
+public:
+    struct Elem {
+        OptType item;
+        mutable size_t count{};
+        [[nodiscard]] bool removed() const { return !item; }
+    };
+    using MutIter = Iter<T, GenericList<T>, typename std::list<Elem>::iterator>;
+    using ConstIter = Iter<const T, const GenericList<T>, typename std::list<Elem>::const_iterator>;
+
+    // Add to the back of the list.
+    T &add_back(T item) { return *list_.emplace_back(Elem{std::move(item), 0}).item; }
+    // Add to the front of the list.
+    T &add_front(T item) { return *list_.emplace_front(Elem{std::move(item), 0}).item; }
+
+    // Remove an item. Removes by comparison to T, so comparing Ts must be defined.
+    // Returns true if item was successfully removed.
+    bool remove(const T &item) {
+        if (auto it = ranges::find(list_, item, &Elem::item); it != list_.end()) {
```

Comment:
I am struggling to find the definition of this flavour of find. The headers are unintelligble and I couldn't find a decent reference page.

Reply:
Right! Yeah, this one's more tricky than it _ought_ to be. The documention is lamentable; I hope it improves as it's now part of the standard (C++20).

Anyway; in general these kinds of range functions take params of the form: `(range, thing_to_do_or_find, [comparison operation in some cases], projection)`. The last few params are optional. In the case of `find`, there's no `comparison operation`, it's always equals. The optional "projection" function takes each element and essentially says "what thing should I actually compare against", the default being "the whole thing". As a special case, the "function" can be a reference to a member variable, or a member function, in which case it's the same as writing a function that returns that variable.

All that to say, we want to find items in the list by their "item" member variable. So we project each `Elem` to its `item` to find by `item`.

The "pass a ref to the field" shortcut is used a bit in the rest of the PR and elsewhere, it's a great way when very commonly you want to find by a field, or sort by a field etc.

Reply:
TLDR find "item" (in this case of type `T`) in the list of `Elem`s by looking at each `Elem`'s `item` field (which is a `T`)